src/spec-lang/tc/semcheck.ts

```diff
@@ -424,3 +429,45 @@ export function scFunctionCall(
 
     throw new Error(`Internal error: NYI semcheck for ${expr.pp()}`);
 }
+
+/**
+ * For Any expression of form old(e1) in `forall(type t in range) e(t)`
+ * throw an error if the expression depends on t.
+ * @param expr
+ * @param ctx
+ * @param typeEnv
+ * @param semMap
+ */
+export function scForAll(expr: SForAll, ctx: SemCtx, typeEnv: TypeEnv, semMap: SemMap): SemInfo {
+    const exprSemInfo = sc(expr.expression, ctx, typeEnv, semMap);
+    const itrSemInfo = sc(expr.iteratorVariable, ctx, typeEnv, semMap);
+    const startSemInfo = sc(expr.start(), ctx, typeEnv, semMap);
+    const endSemInfo = sc(expr.end(), ctx, typeEnv, semMap);
+    const canFail =
+        exprSemInfo.canFail || itrSemInfo.canFail || startSemInfo.canFail || endSemInfo.canFail;
+
+    sc(expr.start(), ctx, typeEnv, semMap);
```

Comment:
We are calling `sc` on `expr.start()` and `expr.end()` twice here. This is unnecessary.